# SLR - BatchGD (FAKE DATA)

In [ ]:
import numpy as np

def simple_linear_regression(X, y, lr=0.01, iter=1000):
    m = len(y)
    theta = np.zeros(2)
    X_b = np.c_[np.ones((m, 1)), X]

    for _ in range(iter):
        preds = X_b.dot(theta)
        errors = preds - y
        gradients = 2/m * X_b.T.dot(errors)
        theta -= lr * gradients
    return theta

np.random.seed(0)
X_simple = 2 * np.random.rand(100, 1)
y_simple = (4 + 3 * X_simple + np.random.randn(100, 1)).flatten()

theta_simple = simple_linear_regression(X_simple, y_simple)
print("Simple:", theta_simple.flatten())

[ 1.35176637 -0.71423072 -0.27905265  1.7228607  -1.30162005 -1.70928477
 -0.99214917  0.36649857  1.25731738 -0.86781457  0.58976316 -1.03582606
 -0.02166085 -0.79006953 -0.16969025 -0.51989903  0.17618246 -1.59928938
  0.05967581 -0.21541292 -1.6965878   1.53434435  1.45711242 -0.7828091
  1.35979365 -1.75701603  0.6002861   0.93402194 -1.73632214 -1.29388153
 -1.68090696 -0.71945626  1.04785377 -1.72345088  0.45468603 -0.61584846
 -0.76064334  0.34161812 -0.42751213 -0.73560674 -0.18978538  1.28603126
 -0.11164022 -1.1397084   0.88116967  0.33623586  0.62181231 -1.66259481
 -0.48564911 -0.22082221  0.95652993 -0.35261891  0.86089434  0.15484172
  0.82250502 -0.48976842 -0.38998731  0.39495216 -0.20937989  1.27971602
  1.67792276 -0.25272047  0.01992961 -0.44836367 -2.1964843  -0.75784049
  1.09940479 -0.9303434   1.50248786  0.64825762  0.25480477 -1.52671619
  0.93131795  1.01304807  0.28504528  0.8501584  -0.9399862   1.26660152
  1.33411671  0.62449024  0.68467778 -1.74289874 -0.

# MLR - BatchGD (FAKE DATA)

In [ ]:
import numpy as np

def multiple_linear_regression(X, y, lr=0.01, iter=1000):
    m = len(y)
    theta = np.zeros(X.shape[1] + 1)
    X_b = np.c_[np.ones((m, 1)), X]

    for _ in range(iter):
        preds = X_b.dot(theta)
        errors = preds - y
        gradients = 2/m * X_b.T.dot(errors)
        theta -= lr * gradients

    return theta

np.random.seed(0)
X_multiple = 2 * np.random.rand(100, 3)
y_multiple = 4 + 5 * X_multiple[:, 0] + 2 * X_multiple[:, 1] + X_multiple[:, 2] + np.random.randn(100)

theta_multiple = multiple_linear_regression(X_multiple, y_multiple)
print("Multiple:", theta_multiple.flatten())

Multiple: [3.66863412 4.85331544 2.0849239  1.23092578]


# SLR - BatchGD (KAGGLE SAMPLE DATASET)

In [ ]:
import numpy as np
import pandas as pd

df = pd.read_csv('/content/sample_data/Salary_dataset.csv')

def simple_linear_regression(X, y, lr=0.01, iter=1000):
    m = len(y)
    theta = np.zeros(2)
    X_b = np.c_[np.ones((m, 1)), X]

    for _ in range(iter):
        preds = X_b.dot(theta)
        errors = preds - y
        gradients = 2/m * X_b.T.dot(errors)
        theta -= lr * gradients

    return theta


X_simple = df['YearsExperience'].values.reshape(-1, 1)
y_simple = df['Salary'].values

theta_simple = simple_linear_regression(X_simple, y_simple)
print("Simple:", theta_simple.flatten())

Simple: [24474.55756611  9504.80132196]



## Comparative Analysis Report: Simple Linear Regression vs Lasso Regularization (Batch Gradient Descent)

In this experiment, a Salary dataset was used to train two models using **Batch Gradient Descent**:

| Method                             | Intercept (θ₀) | Slope (θ₁) |
| ---------------------------------- | -------------- | ---------- |
| **Simple Linear Regression**       | 24474.5575     | 9504.8013  |
| **Lasso Regularization (α = 0.1)** | 24474.5917     | 9504.7949  |

### **Observation**

* The **Lasso model parameters are slightly adjusted** compared to the simple regression model.
* Since the dataset has only **one feature**, the **impact of L1 regularization is minimal**, which is expected because:

  * Lasso (L1) mainly **shrinks coefficients** and can **drive some to zero** in case of multiple features.
  * Here, with just one feature, there’s **no feature selection pressure**, so the effect is subtle.

### **Interpretation**

* The **intercept and slope values are extremely close** in both models.
* This indicates that **overfitting is not a concern** in this simple dataset, and **Lasso does not significantly change the model**.
* However, the **regularized model still adds a penalty**, nudging the slope slightly lower, showing the **soft shrinking behavior** of Lasso.

### **Conclusion**

For **single-feature regression**, adding **L1 regularization has minimal effect** on parameter values. Lasso becomes more meaningful when **multiple features are present**, especially when **sparse feature selection** or **overfitting control** is required.

---

Let me know if you want a **more formal write-up**, maybe in **PDF/report style** with proper headings and interpretation!


# MLR - BatchGD (KAGGLE SAMPLE DATASET)

In [ ]:
import numpy as np
import pandas as pd

df = pd.read_csv('/content/sample_data/Student_Performance.csv')

def multiple_linear_regression(X, y, lr=0.0001, iter=5000):
    m = len(y)
    theta = np.zeros(X.shape[1] + 1)
    X_b = np.c_[np.ones((m, 1)), X]

    for _ in range(iter):
        preds = X_b.dot(theta)
        errors = preds - y
        gradients = 2/m * X_b.T.dot(errors)
        theta -= lr * gradients

    return theta


X_multiple = df[['Hours Studied', 'Previous Scores', 'Sleep Hours', 'Sample Question Papers Practiced']].values
y_multiple = df['Performance Index'].values

theta_multiple = multiple_linear_regression(X_multiple, y_multiple)
print("Multiple:", theta_multiple.flatten())

Multiple: [-0.92086726  2.18894027  0.81830555 -1.45599761 -0.26910558]




## **Comparison: Multiple Linear Regression vs Manual Lasso (Batch GD)**

| Method                             | θ₀ (Intercept) | θ₁     | θ₂     | θ₃      | θ₄      |
| ---------------------------------- | -------------- | ------ | ------ | ------- | ------- |
| **Multiple Linear Regression**     | -0.9208        | 2.1889 | 0.8183 | -1.4559 | -0.2691 |
| **Lasso (Manual Batch GD, α=0.1)** | -0.9225        | 2.1796 | 0.8175 | -1.4440 | -0.2641 |

###**Observation**

* Both models produce **very similar coefficients**, indicating the dataset does **not suffer from multicollinearity or overfitting significantly**.
* The **Lasso version slightly shrinks the weights towards zero**, especially in **θ₁, θ₃, and θ₄**, which shows the **penalty effect of L1 regularization**.
* The **impact is mild**, because **α = 0.1 is small** and the dataset is well-behaved.

### **Conclusion**

* **Multiple Linear Regression** gives pure fitting without penalty.
* **Lasso Regression (Manual)** introduces **controlled shrinkage**, making coefficients slightly smaller.
* While both give almost identical results here, **Lasso becomes more powerful in high-dimensional feature spaces** for **feature selection and overfitting control**.

